ocrd_utils/ocrd_utils/__init__.py

```diff
@@ -547,6 +560,46 @@ def points_from_x0y0x1y1(xyxy):
         x0, y1
     )
 
+def points_of_segment(segment):
+    """Extract the polygon coordinates of a PAGE segment element in page representation
+    
+    Given ``segment``, a PAGE segment object with absolute coordinates
+    (i.e. PageType / RegionType / TextLineType / WordType / GlyphType),
+    return the Coords/@points of that object.
+    
+    (For PageType, either get its Border (if any), or construct a bounding box
+    polygon from its @imageHeight and @imageWidth.)
+    """
+    if segment.__class__.__name__ == 'PageType':
+        if segment.get_Border():
+            return segment.get_Border().get_Coords().points
+        return "%i,%i %i,%i %i,%i %i,%i" % (
+            0,0,
+            0,segment.get_imageHeight(),
+            segment.get_imageWidth(),segment.get_imageHeight(),
+            segment.get_imageWidth(),0)
+    return segment.get_Coords().points
+
+def polygon_for_parent(polygon, parent):
+    """Shrink the given polygon to within the outline of the parent.
+    
+    Given a numpy array ``polygon`` of absolute coordinates, and
+    a PAGE segment object ``parent``, return the intersection of
+    both if ``polygon`` is not fully contained within ``parent``,
+    or ``polygon`` unchanged otherwise.
+    """
+    childp = Polygon(polygon)
+    parentp = Polygon(polygon_from_points(points_of_segment(parent)))
+    if childp.within(parentp):
+        return polygon
+    interp = childp.intersection(parentp)
+    if interp.is_empty:
+        # FIXME: we need a better strategy against this
+        raise Exception("intersection of would-be segment with parent is empty")
```

Comment:
Do we want to be robust against such blatant errors?

Reply:
Yes, probably just show a warning. But we have no context for localizing the message at that point...